File: src/test/java/com/drpicox/game/interpreter/FileTokenizer.java
package com.drpicox.game.interpreter;

import java.util.LinkedList;
import java.util.List;

class FileTokenizer {

    private String[] textLines;
    private int currentIndex;

    FileTokenizer(String file) {
        textLines = file.split("\n");
        currentIndex = 0;
    }

    boolean eof() {
        return currentIndex >= textLines.length;
    }

    int getLineNumber() {
        return currentIndex + 1;
    }

    String match(String regExp) {
        if (this.eof()) return null;
        var currentTextLine = textLines[currentIndex];

        var isMatch = currentTextLine.matches(regExp);
        if (!isMatch) return null;

        return currentTextLine;
    }

    FileTokenizer skipUntil(String regExp) {
        var match = this.match(regExp);
        while (match == null && !this.eof()) {
            currentIndex += 1;
            match = this.match(regExp);
        }
        return this;
    }

    String accept() {
        return this.accept(".*");
    }

    String accept(String regExp) {
        var result = this.match(regExp);
        if (result == null) return null;

        currentIndex += 1;
        return result;
    }

    List<String> acceptUntil(String regExp) {
        var result = new LinkedList<String>();
        var match = this.match(regExp);
        while (match == null && !this.eof()) {
            result.add(textLines[currentIndex]);
            currentIndex += 1;
            match = this.match(regExp);
        }
        return result;
    }
}

File: src/www/__test__/interpreter/FileTokenizer.js
export default class FileTokenizer {
  constructor(file) {
    this._textLines = file.split("\n");
    this._currentIndex = 0;
  }

  eof() {
    return this._currentIndex >= this._textLines.length;
  }

  getLineNumber() {
    return this._currentIndex + 1;
  }

  match(regExp = /.*/) {
    const textLine = this._textLines[this._currentIndex];
    const result = textLine && textLine.match(regExp);
    return result && textLine;
  }

  accept(regExp) {
    const match = this.match(regExp);
    if (match) {
      this._currentIndex += 1;
    }
    return match;
  }

  skipUntil(regExp) {
    let match = this.match(regExp);
    while (!match && !this.eof()) {
      this._currentIndex += 1;
      match = this.match(regExp);
    }
  }

  acceptUntil(regExp) {
    const result = [];
    let match = this.match(regExp);
    while (!match && !this.eof()) {
      result.push(this._textLines[this._currentIndex]);
      this._currentIndex += 1;
      match = this.match(regExp);
    }
    return result;
  }

  toString() {
    const textLines = this._textLines;
    const count = textLines.length;
    const currentIndex = this._currentIndex;
    const result = [];

    let index = 0;
    while (index < currentIndex && index < count) {
      result.push(`  ${this._format(index, textLines)}`);
      index += 1;
    }

    if (index < count) {
      result.push(`> ${this._format(index, textLines)}`);
      index += 1;
    } else {
      result.push(`> EOF`);
    }

    while (index < count) {
      result.push(`  ${this._format(index, textLines)}`);
      index += 1;
    }

    return result.join("\n");
  }

  _format(index, textLines) {
    const lineNumber = index + 1;
    const lineText = textLines[index];
    return `${`    ${lineNumber}`.slice(-5)}: ${lineText}`;
  }
}
